icu4c/source/test/intltest/strtest.cpp

```diff
@@ -265,7 +267,7 @@ StringTest::TestStringPiece() {
         errln("StringPiece() failed");
     }
     // Construct from NULL const char * pointer.
-    StringPiece null(NULL);
+    StringPiece null((const char *)nullptr);
```

Comment:
I think the cast here should be unnecessary.

Reply:
My understanding was that the cast is needed, and that in this case the overload-resolution ambiguity is a feature, not a bug, because we do not want the user casually to express `StringPiece(nullptr)`.  `StringPiece(charptr)` where `charptr == nullptr` shouldn't crash or anything, but a literal `StringPiece(nullptr)` is not one of the intended use-cases.

However, line 555 does say `StringPiece null(nullptr);` and it Just Works... so my understanding must be wrong. Either the cast is unnecessary here, or else it's needed on line 555.

Reply:
I found that we do need a cast for NULL which is implementation-dependent, and got reminded of the same issue in the ICU 59 adoption of char16_t and our adding similar overloads for that. I must have changed this line twice, one to add a cast for NULL where that was needed, and once to replace NULL with nullptr where that was ok for test coverage. Either way, this line continues to test with a null pointer input.

I am pretty sure we want to not break call sites that use a nullptr on purpose but the parameter is actually a StringPiece, so they have been relying on the implicit conversion.